test/e2e/ha_master.go

```diff
@@ -0,0 +1,130 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+	. "github.com/onsi/ginkgo"
+	client "k8s.io/kubernetes/pkg/client/unversioned"
+	"k8s.io/kubernetes/test/e2e/framework"
+	"os/exec"
+	"path"
+	"regexp"
+	"strconv"
+)
+
+func addMasterReplica() error {
+	framework.Logf(fmt.Sprintf("Adding a new master replica:"))
+	v, _, err := framework.RunCmd(path.Join(framework.TestContext.RepoRoot, "hack/e2e-internal/e2e-add-master.sh"))
+	if err != nil {
+		return err
+	}
+	framework.Logf("%q", v)
+	return nil
+}
+
+func removeMasterReplica() error {
+	framework.Logf(fmt.Sprintf("Removing an existing master replica:"))
+	v, _, err := framework.RunCmd(path.Join(framework.TestContext.RepoRoot, "hack/e2e-internal/e2e-remove-master.sh"))
+	if err != nil {
+		return err
+	}
+	framework.Logf("%q", v)
+	return nil
+}
+
+func verifyRCs(c *client.Client, ns string, names []string) {
+	for _, name := range names {
+		framework.ExpectNoError(framework.VerifyPods(c, ns, name, true, 1))
+	}
+}
+
+func createNewRC(c *client.Client, ns string, name string) {
+	_, err := newRCByName(c, ns, name, 1)
+	framework.ExpectNoError(err)
+}
+
+func verifyNumberOfMasterReplicas(expected int) {
+	output, err := exec.Command("gcloud", "compute", "instances", "list",
+		"--project="+framework.TestContext.CloudConfig.ProjectID,
+		"--zone="+framework.TestContext.CloudConfig.Zone,
+		"--regexp="+framework.TestContext.CloudConfig.MasterName + "(-...)?").CombinedOutput()
+	framework.ExpectNoError(err)
+	re := regexp.MustCompile(".*RUNNING")
+	lines := re.FindAllString(string(output), -1)
+	replicas := len(lines)
+	framework.Logf("Num master replicas/expected: %d/%d", replicas, expected)
+	if replicas != expected {
+		framework.Failf("Wrong number of master replicas")
+	}
+}
+
+var _ = framework.KubeDescribe("HA-master [Disruptive]", func() {
+	f := framework.NewDefaultFramework("ha-master")
+	var c *client.Client
+	var ns string
+	var additionalReplicas int
+	var existingRCs []string
+
+	BeforeEach(func() {
+		c = f.Client
+		ns = f.Namespace.Name
+		verifyNumberOfMasterReplicas(1)
+		additionalReplicas = 0
+		existingRCs = make([]string, 0)
+	})
+
+	AfterEach(func() {
+		// Clean-up additional master replicas if the test execution was broken.
+		for i := 0; i < additionalReplicas; i++ {
+			removeMasterReplica()
+		}
+	})
+
+	type Action int
+	const (
+		None Action = iota
+		AddReplica
+		RemoveReplica
+	)
+
+	step := func(action Action) {
+		switch action {
+		case None:
+		case AddReplica:
+			framework.ExpectNoError(addMasterReplica())
+			additionalReplicas++
+		case RemoveReplica:
+			framework.ExpectNoError(removeMasterReplica())
+			additionalReplicas--
+		}
+		verifyNumberOfMasterReplicas(additionalReplicas + 1)
+
+		rcName := "ha-master-" + strconv.Itoa(len(existingRCs))
```

Comment:
Is this step to verify that apiserver works in HA mode? If so, can you please add a comment?